File: PolicyPlus/ListEditor.vb
Public Class ListEditor
    Dim UserProvidesNames As Boolean
    Public FinalData As Object
    Public Function PresentDialog(Title As String, Data As Object, TwoColumn As Boolean) As DialogResult
        UserProvidesNames = TwoColumn
        NameColumn.Visible = TwoColumn
        ElementNameLabel.Text = Title
        EntriesDatagrid.Rows.Clear()
        If Data IsNot Nothing Then
            If TwoColumn Then
                Dim dict As Dictionary(Of String, String) = Data
                For Each kv In dict
                    EntriesDatagrid.Rows.Add(kv.Key, kv.Value)
                Next
            Else
                Dim list As List(Of String) = Data
                For Each entry In list
                    EntriesDatagrid.Rows.Add("", entry)
                Next
            End If
        End If
        FinalData = Nothing
        Return ShowDialog()
    End Function
    Private Sub OkButton_Click(sender As Object, e As EventArgs) Handles OkButton.Click
        If UserProvidesNames Then
            ' Check for duplicate keys
            Dim dict As New Dictionary(Of String, String)
            For Each row As DataGridViewRow In EntriesDatagrid.Rows
                If row.IsNewRow Then Continue For
                Dim key As String = row.Cells(0).Value
                If dict.ContainsKey(key) Then
                    MsgBox("Multiple entries are named """ & key & """! Remove or rename all but one.", MsgBoxStyle.Exclamation)
                    Exit Sub
                Else
                    dict.Add(key, row.Cells(1).Value)
                End If
            Next
            FinalData = dict
        Else
            Dim list As New List(Of String)
            For Each row As DataGridViewRow In EntriesDatagrid.Rows
                If Not row.IsNewRow Then list.Add(row.Cells(1).Value)
            Next
            FinalData = list
        End If
        DialogResult = DialogResult.OK
    End Sub
End Class

File: PolicyPlus/ListEditor.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace PolicyPlus
{
    public partial class ListEditor
    {
        private bool UserProvidesNames;
        public object FinalData;

        public ListEditor()
        {
            InitializeComponent();
        }
        public DialogResult PresentDialog(string Title, object Data, bool TwoColumn)
        {
            UserProvidesNames = TwoColumn;
            NameColumn.Visible = TwoColumn;
            ElementNameLabel.Text = Title;
            EntriesDatagrid.Rows.Clear();
            if (Data is not null)
            {
                if (TwoColumn)
                {
                    Dictionary<string, string> dict = (Dictionary<string, string>)Data;
                    foreach (var kv in dict)
                        EntriesDatagrid.Rows.Add(kv.Key, kv.Value);
                }
                else
                {
                    List<string> list = (List<string>)Data;
                    foreach (var entry in list)
                        EntriesDatagrid.Rows.Add("", entry);
                }
            }
            FinalData = null;
            return ShowDialog();
        }
        private void OkButton_Click(object sender, EventArgs e)
        {
            if (UserProvidesNames)
            {
                // Check for duplicate keys
                var dict = new Dictionary<string, string>();
                foreach (DataGridViewRow row in EntriesDatagrid.Rows)
                {
                    if (row.IsNewRow)
                        continue;
                    string key = Conversions.ToString(row.Cells[0].Value);
                    if (dict.ContainsKey(key))
                    {
                        Interaction.MsgBox("Multiple entries are named \"" + key + "\"! Remove or rename all but one.", MsgBoxStyle.Exclamation);
                        return;
                    }
                    else
                    {
                        dict.Add(key, Conversions.ToString(row.Cells[1].Value));
                    }
                }
                FinalData = dict;
            }
            else
            {
                var list = new List<string>();
                foreach (DataGridViewRow row in EntriesDatagrid.Rows)
                {
                    if (!row.IsNewRow)
                        list.Add(Conversions.ToString(row.Cells[1].Value));
                }
                FinalData = list;
            }
            DialogResult = DialogResult.OK;
        }
    }
}